File: ios/Sources/VoiceRecorder/Service/VoiceRecorderService.swift
import Foundation

/// Service layer that orchestrates recording operations.
final class VoiceRecorderService {
    /// Platform adapter for device and file operations.
    private let platform: RecorderPlatform
    /// Closure used to check microphone permission.
    private let permissionChecker: () -> Bool
    /// Active recorder instance for the current session.
    private var recorder: RecorderAdapter?

    init(platform: RecorderPlatform, permissionChecker: @escaping () -> Bool) {
        self.platform = platform
        self.permissionChecker = permissionChecker
    }

    /// Returns whether the device can record audio.
    func canDeviceVoiceRecord() -> Bool {
        return platform.canDeviceVoiceRecord()
    }

    /// Returns whether the app has microphone permission.
    func hasAudioRecordingPermission() -> Bool {
        return permissionChecker()
    }

    /// Starts a recording session or throws a service error.
    func startRecording(
        options: RecordOptions?,
        onInterruptionBegan: @escaping () -> Void,
        onInterruptionEnded: @escaping () -> Void
    ) throws {
        if !platform.canDeviceVoiceRecord() {
            throw VoiceRecorderServiceError(code: ErrorCodes.deviceCannotVoiceRecord)
        }

        if !permissionChecker() {
            throw VoiceRecorderServiceError(code: ErrorCodes.missingPermission)
        }

        if recorder != nil {
            throw VoiceRecorderServiceError(code: ErrorCodes.alreadyRecording)
        }

        let nextRecorder = CustomMediaRecorder()
        nextRecorder.onInterruptionBegan = onInterruptionBegan
        nextRecorder.onInterruptionEnded = onInterruptionEnded
        let started = nextRecorder.startRecording(recordOptions: options)
        if !started {
            recorder = nil
            throw VoiceRecorderServiceError(code: ErrorCodes.deviceCannotVoiceRecord)
        }

        recorder = nextRecorder
    }

    /// Stops recording and returns the payload asynchronously.
    func stopRecording(completion: @escaping (Result<RecordData, VoiceRecorderServiceError>) -> Void) {
        guard let recorder = recorder else {
            completion(.failure(VoiceRecorderServiceError(code: ErrorCodes.recordingHasNotStarted)))
            return
        }

        recorder.stopRecording { [weak self] stopSuccess in
            guard let self = self else {
                completion(.failure(VoiceRecorderServiceError(code: ErrorCodes.failedToFetchRecording)))
                return
            }

            if !stopSuccess {
                self.recorder = nil
                completion(.failure(VoiceRecorderServiceError(code: ErrorCodes.failedToMergeRecording)))
                return
            }

            let audioFileUrl = recorder.getOutputFile()
            let fileExtension = audioFileUrl.pathExtension.lowercased()
            let mimeType = fileExtension == "m4a" ? "audio/mp4" : "audio/aac"
            let sendDataAsBase64 = recorder.options?.directory == nil
            let recordDataBase64 = sendDataAsBase64 ? self.platform.readFileAsBase64(audioFileUrl) : nil
            let uri = sendDataAsBase64 ? nil : audioFileUrl.path
            let recordData = RecordData(
                recordDataBase64: recordDataBase64,
                mimeType: mimeType,
                msDuration: self.platform.getDurationMs(audioFileUrl),
                uri: uri
            )

            self.recorder = nil
            if (sendDataAsBase64 && recordData.recordDataBase64 == nil) || recordData.msDuration < 0 {
                completion(.failure(VoiceRecorderServiceError(code: ErrorCodes.emptyRecording)))
            } else {
                completion(.success(recordData))
            }
        }
    }

    /// Pauses the active recording session.
    func pauseRecording() throws -> Bool {
        guard let recorder = recorder else {
            throw VoiceRecorderServiceError(code: ErrorCodes.recordingHasNotStarted)
        }
        return recorder.pauseRecording()
    }

    /// Resumes a paused recording session.
    func resumeRecording() throws -> Bool {
        guard let recorder = recorder else {
            throw VoiceRecorderServiceError(code: ErrorCodes.recordingHasNotStarted)
        }
        return recorder.resumeRecording()
    }

    /// Returns the current recording status.
    func getCurrentStatus() -> CurrentRecordingStatus {
        guard let recorder = recorder else {
            return .NONE
        }
        return recorder.getCurrentStatus()
    }
}

File: android/src/main/java/app/independo/capacitorvoicerecorder/service/VoiceRecorderService.java
package app.independo.capacitorvoicerecorder.service;

import app.independo.capacitorvoicerecorder.adapters.PermissionChecker;
import app.independo.capacitorvoicerecorder.adapters.RecorderAdapter;
import app.independo.capacitorvoicerecorder.adapters.RecorderPlatform;
import app.independo.capacitorvoicerecorder.core.CurrentRecordingStatus;
import app.independo.capacitorvoicerecorder.core.ErrorCodes;
import app.independo.capacitorvoicerecorder.core.RecordData;
import app.independo.capacitorvoicerecorder.core.RecordOptions;
import app.independo.capacitorvoicerecorder.platform.NotSupportedOsVersion;
import java.io.File;
import java.util.function.Consumer;

/** Service layer that orchestrates recording operations. */
public class VoiceRecorderService {

    /** Platform adapter that owns file and recorder creation. */
    private final RecorderPlatform platform;
    /** Permission checker injected from the bridge layer. */
    private final PermissionChecker permissionChecker;
    /** Current recorder instance for an active session. */
    private RecorderAdapter recorder;

    public VoiceRecorderService(RecorderPlatform platform, PermissionChecker permissionChecker) {
        this.platform = platform;
        this.permissionChecker = permissionChecker;
    }

    /** Returns whether the device can record audio. */
    public boolean canDeviceVoiceRecord() {
        return platform.canDeviceVoiceRecord();
    }

    /** Returns whether the app has microphone permission. */
    public boolean hasAudioRecordingPermission() {
        return permissionChecker.hasAudioPermission();
    }

    /** Starts a recording session or throws a service exception. */
    public void startRecording(
        RecordOptions options,
        Runnable onInterruptionBegan,
        Runnable onInterruptionEnded,
        Consumer<Float> onVolumeChanged
    ) throws VoiceRecorderServiceException {
        if (!platform.canDeviceVoiceRecord()) {
            throw new VoiceRecorderServiceException(ErrorCodes.DEVICE_CANNOT_VOICE_RECORD);
        }

        if (!permissionChecker.hasAudioPermission()) {
            throw new VoiceRecorderServiceException(ErrorCodes.MISSING_PERMISSION);
        }

        if (platform.isMicrophoneOccupied()) {
            throw new VoiceRecorderServiceException(ErrorCodes.MICROPHONE_BEING_USED);
        }

        if (recorder != null) {
            throw new VoiceRecorderServiceException(ErrorCodes.ALREADY_RECORDING);
        }

        try {
            recorder = platform.createRecorder(options);
            recorder.setOnInterruptionBegan(onInterruptionBegan);
            recorder.setOnInterruptionEnded(onInterruptionEnded);
            recorder.setOnVolumeChanged(onVolumeChanged);
            recorder.startRecording();
        } catch (Exception exp) {
            recorder = null;
            throw new VoiceRecorderServiceException(ErrorCodes.FAILED_TO_RECORD, exp);
        }
    }

    /** Stops the active recording session and returns the payload. */
    public RecordData stopRecording() throws VoiceRecorderServiceException {
        if (recorder == null) {
            throw new VoiceRecorderServiceException(ErrorCodes.RECORDING_HAS_NOT_STARTED);
        }

        RecordOptions options = recorder.getRecordOptions();

        try {
            recorder.stopRecording();
            File recordedFile = recorder.getOutputFile();
            if (recordedFile == null) {
                throw new VoiceRecorderServiceException(ErrorCodes.FAILED_TO_FETCH_RECORDING);
            }

            String recordDataBase64 = null;
            String uri = null;
            if (options.directory() != null) {
                uri = platform.toUri(recordedFile);
            } else {
                recordDataBase64 = platform.readFileAsBase64(recordedFile);
            }

            int duration = platform.getDurationMs(recordedFile);
            RecordData recordData = new RecordData(recordDataBase64, duration, "audio/aac", uri);
            if ((recordDataBase64 == null && uri == null) || recordData.getMsDuration() < 0) {
                throw new VoiceRecorderServiceException(ErrorCodes.EMPTY_RECORDING);
            }

            return recordData;
        } catch (VoiceRecorderServiceException exp) {
            throw exp;
        } catch (Exception exp) {
            throw new VoiceRecorderServiceException(ErrorCodes.FAILED_TO_FETCH_RECORDING, exp);
        } finally {
            if (options.directory() == null && recorder != null) {
                recorder.deleteOutputFile();
            }
            recorder = null;
        }
    }

    /** Pauses the active recording session. */
    public boolean pauseRecording() throws VoiceRecorderServiceException {
        if (recorder == null) {
            throw new VoiceRecorderServiceException(ErrorCodes.RECORDING_HAS_NOT_STARTED);
        }
        try {
            return recorder.pauseRecording();
        } catch (NotSupportedOsVersion exception) {
            throw new VoiceRecorderServiceException(ErrorCodes.NOT_SUPPORTED_OS_VERSION, exception);
        }
    }

    /** Resumes a paused recording session. */
    public boolean resumeRecording() throws VoiceRecorderServiceException {
        if (recorder == null) {
            throw new VoiceRecorderServiceException(ErrorCodes.RECORDING_HAS_NOT_STARTED);
        }
        try {
            return recorder.resumeRecording();
        } catch (NotSupportedOsVersion exception) {
            throw new VoiceRecorderServiceException(ErrorCodes.NOT_SUPPORTED_OS_VERSION, exception);
        }
    }

    /** Returns the current recording status. */
    public CurrentRecordingStatus getCurrentStatus() {
        if (recorder == null) {
            return CurrentRecordingStatus.NONE;
        }
        return recorder.getCurrentStatus();
    }
}
